0015-automatic-cluster-upgrades/0015-automatic-cluster-upgrades.md

```diff
@@ -0,0 +1,151 @@
+
+## Current state
+
+Right now the automatic upgrades are carried by Solution Engineers agreeing with the customer on upgrade windows and environments selection. The idea is to automate all this behavior and information under a Kubernetes operator(s) and Custom Resource(s).
+
+The Solution engineer plans cluster upgrades based on cluster type (environment, team,...) and maintenance windows offered by the customers. The upgrade is just a change on the cluster version field on the cluster Custom Resource (CR).
+
+Based on the fact Giant Swarm is moving to Cluster API this is story let us define a new greenfield scenario where we can design a model that can be more generic and not focus only in Giant Swarm platform. The idea would be contribute as much as possible with upstream due the fact most of this functionality would be useful for everyone that manages clusters.
+
+## Concepts
+
+### Cluster upgrade entity
+
+There are different reason why we are interested in having an object to define the upgrade intent. 
+
+- As we define in the user stories, we/customers would like to schedule upgrades on certain date from running version to a specific one. That way customer can plan upgrades ahead and so do we.
+
+- We have a registry of upgrades done to a cluster for auditing and tracking purposes.
+
+- We can save cluster upgrade status displaying if something goes wrong or not.
+
+- In theory we can include apps upgrades together or separately.
+
+### Maintenance schedule entity
+
+Our customer can have a different set of constraints when it comes to allow changes in their clusters. The maintenance schedule would define those and then the clusters can rely on a policy to enforce the desired behavior.
+
+The maintenance schedule would be possible to be referenced from different cluster resources (Cluster,Machine Pool, KubeadmControlPlane, Apps,...) letting controllers to decide if the resource would be or not reconciled.
+
+The maintenance schedule would contain
+
+- A way to define dates or times where clusters (or relatives) can be modified.
+
+- A way to define dates or times where clusters (or relatives) cannot be modified.
+
+## Technical aspects
+
+We envision aforementioned concepts would become custom resource definitions and there will be additional controllers that would enhance and ensure the correct behavior.
+
+### Resources
+
+#### ClusterUpgrade
+
+Example resource:
+
+```yaml
+apiVersion: 
+kind: ClusterUpgrade
+metadata:
+  name: upgrade-to-release-v20
+  namespace: default
+spec:
+  clusterRef:
+    apiVersion: cluster.x-k8s.io/v1alpha3
+    kind: Cluster
+    name: foo01
+  releaseRef:
+    apiVersion: release.giantswarm.io/v1alpha1
+    kind: Release
+    name: v13.0.1
+  startTime: 2021-03-10T07:00:00Z
```

Comment:
Thanks to @mig4 to clarify `upgradeAfter` has other meaning https://github.com/kubernetes-sigs/cluster-api/pull/4285/files. 

Anyway we could think in two different approaches that can live together. One is to add a field like `startTime` to define exactly when cluster must be upgraded (even not upgrade the cluster if there has been a problem as X minutes has passed the startTime), and/or add something like `upgradeAfter` which just means you want to upgrade when the time goes over the `upgradeAfter` field. I tend to think to avoid later, first, the field already exist upstream with a different purpose and second, this sounds like a policy/maintananceWindow scope, WDYT?

 @mig4 @fiunchinho

Reply:
If we have the `MaintenanceSchedule` CRD to define the maintenance window, I guess the `ClusterUpgrade` would hold a reference to a `MaintenanceSchedule`?

Reply:
hmm, maybe that is in the status then?

Reply:
How come? wouldn't a `ClusterUpgrade` need to select which `MaintenanceSchedule` to use?

Reply:
hmm, seems more complex now that I think of it, the schedule only gives permission to upgrade right? then an upgrade scheduler can use this to create upgrade CRs automatically, so there should not be a need for any reference I think. But then my question would be, why even need a startTime here? If you want to trigger an upgrade just create the CR, start time is the creation time of the CR, the creator of the CR (here the cluster upgrade scheduler) has to adhere to maintenanceschedule or it could be validated in admission. In the end it depends a lot on how you want to implement this in the end, but as it is described below to me it makes more sense if the scheduler knows about maintenance and then the executor just executes.

From what we tried to come up with in batman back then for auto-updates of apps, we thought to align it more with the concepts that people are used to from autoscaling in HPA as in the end the upgrade is just a scale operation on a version field, so you could set an autoupgrade policy like upgrade all patch or all patch and minor (that can be expressed with semver notation even) and reference the maintenance schedule in or from there, then with simple CRs like the App CR the upgrade controller can just increment the version field based on a new version being present and in more complex situations like the cluster upgrade you might wanna delegate to another controller that can handle upgrade strategies, and delegation could be through sth like the upgrade CR here.

The above is not very far from what you describe here. It only is missing autoupgrade policy being configurable and upgrade strategies, but that's ok if you wanna define an MVP scope here and add stuff like that later.

Reply:
> But then my question would be, why even need a startTime here? 

I guess we will enable teams to plan upgrades. So they say next Monday we will upgrade cluster X and cluster Y at 9:00 (and they prepare the objects), on Tuesday afternoon we will upgrade cluster Z...

> From what we tried to come up with in batman back then for auto-updates of apps, we thought to align it more...

I started with that idea but in reality customers (at least as it is today) dont like to lose the control of an upgrade. It could be interesting to get all reasons why and maybe build functionality on the upgrade mechanism to avoid them